### PR TITLE
Batch OT losses (Sinkhorn + Gromov)

### PR DESCRIPTION
**Pull request by KrzakalaPaul**
## Types of changes

Add the ot.batch module for solving N optimal transport problems at the same time in parallel. The two main features are:

- ot.batch.solve_batch, the API is the same as for ot.solve (with less features for now).
- ot.batch.solve_gromov_batch, the API is the same as for ot.solve_gromov (with less features for now).

Some examples have been added to examples/batch

## Motivation and context / Related issue

It is often the case that one needs to solve N optimal transport problem at the same time. At the moment, the only way to do this in POT was using a for loop. This is very inefficient for data stored on a GPU for instance. 
Instead, it is now possible to use ot.batch for solving N problems at the same time with fully batch parallel operations. As demonstrated in examples/batch/demo_efficiency.py this can lead to a speed of 1 or 2 orders of magnitude. 

## How has this been tested (if it applies)

New tests have been added to test/batch/ to ensure that using ot.solve_batch yields the same results that using a for loops of ot.solve. Idem for ot.solve_batch_gromov. 

## PR checklist

- [x] I have read the [**CONTRIBUTING**](https://pythonot.github.io/contributing.html) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: test/test_bregman.py [hunk 1]
```diff
@@ -1,9 +1,6 @@
-"""Tests for module bregman on OT with bregman projections"""
```
Comment (rflamary) [inline test/test_bregman.py]: This is a bad modification, those are the existing tests no?

File: ot/batch/_linear.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    M,
+    a=None,
+    b=None,
+    epsilon=1e-3,
+    max_iter=1000,
+    tol=1e-5,
+    log_dual=True,
+    grad="detach",
+):
+    r"""Solves the linear optimal transport problem using Bregman projections.
```
Comment (rflamary) [inline ot/batch/_linear.py]: Add entropic here this is not exact OT

File: ot/batch/_linear.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    >>> Y = np.random.rand(B, nt, d)
+    >>> M = cost_matrix_l2_batch(X, Y)
+    >>> res = linear_solver_batch(M=M, epsilon=0.01, a=a, b=b)
+    >>> res.plan.shape
+    (B, ns, nt)
+    >>> res.value.shape
+    (B,)
+    """
+
+    nx = get_backend(a, b, M)
```
Comment (rflamary) [inline ot/batch/_linear.py:308]: raise error if reg=0, also maybe make reg mandatory (to be clear thta this is entropic OT?)

File: ot/batch/_linear.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        Regularization parameter
+    a : array-like, shape (B, ns)
+        Source distribution (optional). If None, uniform distribution is used.
+    b : array-like, shape (B, nt)
+        Target distribution (optional). If None, uniform distribution is used.
+    max_iter : int
+        Maximum number of iterations
+    tol : float
+        Tolerance for convergence
+    log_dual : bool
```
Comment (rflamary) [inline ot/batch/_linear.py]: use same parameter solver as ot.solve : solver='log_sinkhorn'

File: docs/source/user_guide.rst [hunk 5]
```diff
@@ -1217,3 +1217,6 @@ References
     couplings <http://proceedings.mlr.press/v89/forrow19a/forrow19a.pdf>`_. In
     The 22nd International Conference on Artificial Intelligence and Statistics
     (pp. 2454-2465). PMLR.
+
+.. [41] Xu, H., Luo, D., & Carin, L. (2019). `Scalable Gromov-Wasserstein learning for graph partitioning and matching
```
Comment (rflamary) [inline docs/source/user_guide.rst:1221]: add to readme file not necessarilily here

File: examples/batch/README.txt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+
+
+Batch parrallel optimal transport
```
Comment (rflamary) [inline examples/batch/README.txt]: mopve examples in backend folder

File: examples/batch/demo_pitfalls.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    is_pos = (loss >= -1e-6).all()
+    print("All losses are positive: ", is_pos)
+
+    print("")
+    print(
+        "Conclusion: when T does not satisfy marginals, it is important to set recompute_const=True to ensure that the loss computation is correct."
+    )
+
+
+if __name__ == "__main__":
```
Comment (rflamary) [inline examples/batch/demo_pitfalls.py]: no if name in exmaples

File: examples/batch/demo_pitfalls.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    print(f"... solver took: {end - start:.4f} seconds")
+    print(f"... final value: {res.value.mean().item():.4f}")
+
+    print("")
+    print(
+        "Conclusion: symmetric=False is slightly faster but leads to suboptimal values because some incorrect approximations are made."
+    )
+
+
+def demo_recompute_const():
```
Comment (rflamary) [inline examples/batch/demo_pitfalls.py]: no function, use section separated by # %%

File: examples/batch/demo_efficiency.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
```
Comment (rflamary) [inline examples/batch/demo_efficiency.py]: use rst formating and sections, also factorize in lass examples+ rename as plot_* instead of demo

File: ot/backend.py [hunk 10]
```diff
@@ -1195,6 +1219,9 @@ def exp(self, a):
     def log(self, a):
         return np.log(a)
 
+    def logsumexp(self, a, axis=None, keepdims=False):
```
Comment (rflamary) [inline ot/backend.py:1208]: delete all logsumexp!!!!!!! Alerday there

Comment (rflamary) [reply ot/backend.py:1208]: ok...moving is is OK

File: ot/backend.py [hunk 11]
```diff
@@ -1543,6 +1570,12 @@ def set_gradients(self, val, inputs, grads):
     def _detach(self, a):
         return jax.lax.stop_gradient(a)
 
+    @contextmanager
+    def set_grad_enabled(self, enabled):
+        # No global flag in JAX, gradients are enabled/disabled by using
```
Comment (rflamary) [inline ot/backend.py]: contetx manager needs to work on all backends...

File: ot/batch/__init__.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+Batch operations for optimal transport.
+"""
+
+# Author: Remi Flamary <remi.flamary@unice.fr>
+#         Paul Krzakala <paul.krzakala@gmail.com>
+#
+# License: MIT License
+
+from ._linear import solve_batch
+from ._quadratic import solve_gromov_batch
```
Comment (rflamary) [inline ot/batch/__init__.py]: add losse sin import and also add exmaples ;)

File: ot/batch/_linear.py [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    -------
+    M : array-like, shape (B, ns, nt)
+        Cost matrix where M[bij] is the cost between sample i in batch b and sample j in batch b.
+    """
+
+    if nx is None:
+        nx = get_backend(X, Y)
+    entr_y = nx.sum(Y * nx.log(Y + 1e-10), axis=-1)  # B x m
+    if logits_X:
+        M = entr_y[:, None, :] - Y[:, None, :] + nx.log(X + 1e-10)[:, :, None]
```
Comment (rflamary) [inline ot/batch/_linear.py]: mayeb paramerize eps

File: ot/batch/_linear.py [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+    else:
+        raise ValueError(f"Unknown metric: {metric}")
+    return loss_linear_batch(M, T)
+
+
+def entropy_batch(T, nx=None):
+    """Computes the entropy of the transport plan T."""
+    if nx is None:
+        nx = get_backend(T)
+    return -nx.sum(T * nx.log(T + 1e-10), axis=(1, 2))
```
Comment (rflamary) [inline ot/batch/_linear.py]: parameter eps

File: ot/batch/_linear.py [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+        - res.value_linear : Linear OT loss with the optimal OT plan
+
+        See :any:`OTResult` for more information.
+    """
+
+    nx = get_backend(a, b, M)
+
+    if log_dual:
+        K = -M / epsilon
+        out = bregman_log_batch(K, a, b, nx=nx, max_iter=max_iter, tol=tol, grad=grad)
```
Comment (rflamary) [inline ot/batch/_linear.py]: renem bregman_proj_log

File: ot/batch/_linear.py [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+
+    entr = entropy_batch(T, nx=nx)
+    value_linear = loss_linear_batch(M, T)
+    value = value_linear + epsilon * entr
+    log = {"n_iter": out["n_iters"]}
+
+    res = OTResult(
+        value=value,
+        value_linear=value_linear,
+        plan=T,
```
Comment (rflamary) [inline ot/batch/_linear.py:352]: add dual potentials

File: ot/batch/_linear.py [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+#         Paul Krzakala <paul.krzakala@gmail.com>
+#
+# License: MIT License
+
+from ..backend import get_backend
+from ..utils import OTResult
+from ._utils import bregman_log_batch, bregman_batch
+
+
+def cost_matrix_lp_batch(X, Y, p=2, q=1, nx=None):
```
Comment (rflamary) [inline ot/batch/_linear.py]: dist instead of cost matrix

File: ot/batch/_linear.py [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+    if nx is None:
+        nx = get_backend(X, Y)
+    M = nx.abs(X[:, :, None] - Y[:, None, :]) ** p
+    M = M.sum(axis=-1)
+    if q != p:
+        M = M ** (q / p)
+    return M
+
+
+def cost_matrix_l2_batch(X, Y, squared=True, nx=None):
```
Comment (rflamary) [inline ot/batch/_linear.py]: same

File: ot/batch/_linear.py [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+    XX = nx.sum(X**2, axis=-1, keepdims=True)
+    YY = nx.sum(Y**2, axis=-1, keepdims=True)
+    YY = nx.transpose(YY, axes=(0, 2, 1))
+    M = XX + YY - 2 * nx.einsum("bid,bjd->bij", X, Y)
+    if not squared:
+        M = nx.sqrt(M)
+    return M
+
+
+def cost_matrix_kl_batch(X, Y, logits_X=False, nx=None):
```
Comment (rflamary) [inline ot/batch/_linear.py]: same

File: ot/batch/_quadratic.py [hunk 20]
```diff
@@ -0,0 +1,480 @@
...
+from abc import ABC, abstractmethod
+
+
+def transpose(C, nx=None):
+    if nx is None:
+        nx = get_backend(C)
+    return nx.transpose(C, (0, 2, 1)) if C.ndim == 3 else nx.transpose(C, (0, 2, 1, 3))
+
+
+class QuadraticMetric(ABC):
```
Comment (rflamary) [inline ot/batch/_quadratic.py]: PairwiseMetric instead of quadratic

File: ot/batch/_quadratic.py [hunk 21]
```diff
@@ -0,0 +1,480 @@
...
+
+    def h1(self, C1, nx=None):
+        return C1 if self.logits else nx.log(C1 + 1e-15)
+
+    def h2(self, C2, nx=None):
+        return C2
+
+
+def get_quadratic_metric(metric_name):
+    if metric_name == "L2":
```
Comment (rflamary) [inline ot/batch/_quadratic.py]: .upper()

File: ot/batch/_quadratic.py [hunk 22]
```diff
@@ -0,0 +1,480 @@
...
+    if not symmetric:
+        dot_t = nx.einsum("bijd,bjk->bikd", transpose(hC1), T)
+        dot_t = nx.einsum("bikd,bjkd->bijd", dot_t, transpose(hC2))
+        dot_t = dot_t.sum(axis=-1)
+        dot = (dot + dot_t) / 2  # Average the two symmetric terms
+
+    return const - dot
+
+
+def loss_quadratic_batch(L, T, recompute_const=False, symmetric=True, nx=None):
```
Comment (rflamary) [inline ot/batch/_quadratic.py:155]: loss_gromov_batch

File: ot/batch/_quadratic.py [hunk 23]
```diff
@@ -0,0 +1,480 @@
...
+def solve_gromov_batch(
+    C1,
+    C2,
+    a=None,
+    b=None,
+    loss="L2",
+    symmetric=None,
+    M=None,
+    alpha=None,
+    epsilon=1e-2,
```
Comment (rflamary) [inline ot/batch/_quadratic.py]: reg + mandatory

File: ot/batch/_linear.py [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+    if nx is None:
+        nx = get_backend(T)
+    return -nx.sum(T * nx.log(T + 1e-10), axis=(1, 2))
+
+
+def solve_batch(
+    M,
+    a=None,
+    b=None,
+    epsilon=1e-3,
```
Comment (rflamary) [inline ot/batch/_linear.py]: reg + mandatory

File: ot/batch/_quadratic.py (hunk 23 above)
Comment (rflamary) [reply ot/batch/_quadratic.py]: call it reg_inner

File: ot/batch/_quadratic.py [hunk 25]
```diff
@@ -0,0 +1,480 @@
...
+        L, T, nx=nx, recompute_const=True, symmetric=symmetric
+    )  # Always recompute const for accurate value
+    value = (1 - alpha) * value_linear + alpha * value_quadratic
+    log = {"n_iter": iter}
+
+    res = OTResult(
+        value=value,
+        value_linear=value_linear
+        * (1 - alpha),  # Weight the linear value for consistency with ot.solve_gromov
+        value_quad=value_quadratic * alpha,  # idem
```
Comment (rflamary) [inline ot/batch/_quadratic.py:491]: potentials

File: ot/batch/_utils.py [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+
+
+def bop(a, b, nx):
+    """
+    Batched outer product for vectors a and b.
+    """
+    return nx.einsum("bi,bj->bij", a, b)
+
+
+def bregman_batch(K, a=None, b=None, nx=None, max_iter=10000, tol=1e-5, grad="detach"):
```
Comment (rflamary) [inline ot/batch/_utils.py]: bregman_proj_batch

File: ot/batch/_quadratic.py [hunk 27]
```diff
@@ -0,0 +1,606 @@
...
+    loss="sqeuclidean",
+    symmetric=None,
+    M=None,
+    alpha=None,
+    T_init=None,
+    max_iter=50,
+    tol=1e-5,
+    max_iter_inner=50,
+    tol_inner=1e-5,
+    grad="detach",
```
Comment (rflamary) [inline ot/batch/_quadratic.py]: we need a gradient by default

File: test/batch/test_solve_batch.py [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+
+    res = solve_batch(
+        M,
+        a=None,
+        b=None,
+        reg=reg,
+        max_iter=max_iter,
+        tol=tol,
+        solver="log_sinkhorn",
+        grad="detach",
```
Comment (rflamary) [inline test/batch/test_solve_batch.py:49]: gradient by default (at elast envelop)
